astropy/coordinates/builtin_frames/galactocentric.py

```diff
@@ -30,6 +35,47 @@
 _ROLL0 = Angle(58.5986320306*u.degree)
 
 
+class _StateProxy(MappingView):
+    """
+    `~collections.abc.MappingView` with a read-only ``getitem`` through
+    `~types.MappingProxyType`.
+
+    """
+
+    def __init__(self, mapping):
+        super().__init__(mapping)
+        self._mappingproxy = MappingProxyType(self._mapping)  # read-only
+
+    def __getitem__(self, key):
+        """Read-only ``getitem``."""
+        return self._mappingproxy[key]
+
+    def copy(self):
+        """A mutable dictionary deepcopy."""
+        return copy.deepcopy(self)
+
+
+def _reduce_stateproxy(sp: _StateProxy):
```

Comment:
I think the advantage of the class above is that it can now define `__getstate__` instead of registering this override. Or am I missing something? I am actually slightly confused why this needed - why not adjust the `.copy` method itself?

Reply:
Unfortunately, a `__getstate__` method doesn't work in this case. I could create a custom ``__deepcopy__`` method, but the `copyreg` method also allows the object to be pickled.

Reply:
https://docs.python.org/3/library/copy.html
ultimate and penultimate paragraphs

Reply:
But with this code, doesn't it mean that if you unpickle the pickled instance, you get a `dict` rather than a `_StateProxy`?

I'm actually not quite sure whether it is possible to pickle and unpickle the class itself, but it would not seem good that if I pickled the registry for use in another process, all the registries would become `dict`... Especially if the goal is to ensure to avoid a place being made to propagate back.

Overall, I do wonder if we're not over-engineering this. Might it not make more sense to just create a new dict (i.e., do the deep copy) inside the `get_registry` method? (I vaguely fear this is returning to a previous design -- if so, sorry!).

Reply:
If I switch from `copyreg` to implementing `__deepcopy__` in `_StateProxy`, then the deep copy will happen inside the `get_from_registry` method and it will have identical behavior to `MappinProxy` (ie read-only), but also implement `getitem`. It will not be able to be pickled, so should not have any unexpected behavior.